pennylane_lightning/src/Apply.cpp

```diff
@@ -106,3 +112,101 @@ void Pennylane::apply(
     }
 
 }
+
+void Pennylane::adjointJacobian(
+    StateVector& phi,
+    double* jac,
+    const vector<string>& observables,
+    const vector<vector<double> >& obsParams,
+    const vector<vector<unsigned int> >& obsWires,
+    const vector<string>& operations,
+    const vector<vector<double> >& opParams,
+    const vector<vector<unsigned int> >& opWires,
+    const vector<int>& trainableParams,
+    int paramNumber
+) {
+    vector<Pennylane::StateVector> lambdas;
+    size_t numObservables = observables.size();
+    size_t trainableParamNumber = trainableParams.size() - 1;
+
+    for (unsigned int i = 0; i < numObservables; i++) {
+        // copy |phi> and apply observables one at a time
+        CplxType* phiCopyArr = new CplxType[phi.length];
+        std::memcpy(phiCopyArr, phi.arr, sizeof(CplxType)*phi.length);
+        Pennylane::StateVector phiCopy(phiCopyArr, phi.length);
+
+        Pennylane::constructAndApplyOperation(
+            phiCopy,
+            observables[i],
+            obsWires[i],
+            obsParams[i],
+            false,
+            obsWires[i].size()
+        );
+        lambdas.push_back(phiCopy);
+    }
+
+    for (int i = operations.size() - 1; i >= 0; i--) {
+        if (opParams[i].size() > 1) {
+            throw std::invalid_argument("The operation is not supported using the adjoint differentiation method");
+        } else if ((operations[i] != "QubitStateVector") && (operations[i] != "BasisState")) {
+            // copy |phi> to |mu> before applying Uj*
+            CplxType* phiCopyArr = new CplxType[phi.length];
```

Comment:
Could you just allocate the memory once and then always copy the `phi` statevector there, overwriting it's value from the last loop?